### ISSUE TITLE
enum support in base

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#3080]
Currently [enum.jl lives in extras](https://github.com/JuliaLang/julia/blob/master/extras/enum.jl). Since enums are generally considered useful, we should do a proper design pass on enums and add them to base. Part of #2422. For design inspiration, the PEP might be helpful: http://www.python.org/dev/peps/pep-0435/.

**Comment by quinnj** [comment]
Having advocated for this, I'm taking a stab at implementing enums (based heavily on @vtjnash's work in #2988). I've also been reviewing Python's internal implementation of enum.Enum. 
One big design question that hasn't really been addressed is if we would want Enum values to be restricted to the initial enumeration. Python does enforce this and uses a set internally to track. Much of the discussion and Jameson's initial work is based much more simply on a single Enum + `@enum` macro that doesn't enforce. This would allow something like the following
```julia
@enum FRUIT apple=1, orange=2, kiwi=3

f(x::FRUIT) = "hey, i'm a fruit"

f(FRUIT(4)) # valid, even though we didn't define a FRUIT(4)
```
Thoughts on this?

**Comment by eschnett** [comment]
This should be invalid. I like to use enums to ensure all cases are handled, and if there is an easy way to accidentally create invalid enums, this is thwarted.

**Comment by elextr** [comment]
Is this valid?

    @enum FRUIT apple=1, orange=2, kiwi=4

    fruit_salad = FRUIT(6)

even though 6 is not one of the enumeration values?

**Comment by catawbasam** [comment]
I agree with @eschnett.  FRUIT(6) should be invalid.

**Comment by elextr** [comment]
So why give them numbers and convert to/from int?  Don't leak the implementation :)

The sets of bits passed to/from C functions in ints are different things to enums as @eschnett wants to use them, they shouldn't be the same concept.  

PS I'm not advocating for/against, just asking some questions to ensure that they are considered.

**Comment by ivarne** [comment]
The benefit of leaking the implementation is that it gives a direct way to translate things to C (and other languages, or storage systems that use integers for categorical values). It's also a great way to name some magic numbers that is used in calculations (eg. bitmasks, but arguably that could be a different concept from a enum).

I just came to think of `Char` as a special case of enum, but 1114112 unicode characters might require more infrastructure than what we generally want for a 2-20 element enum.

**Comment by quinnj** [comment]
Closed by #10168
